src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py

```diff
@@ -0,0 +1,326 @@
+        code.
+
+        :param node: a Sum, Minval or Maxval intrinsic.
+        :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+        :param options: options for the transformation.
+        :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node)
+
+        array_ref, dimension_ref, mask_ref = self._get_args(node)
+
+        # Determine the literal value of the dimension argument
+        dimension_literal = None
+        if not dimension_ref:
+            # there is no dimension argument
+            pass
+        elif isinstance(dimension_ref, Literal):
+            dimension_literal = dimension_ref
+        elif (isinstance(dimension_ref, Reference) and
+              dimension_ref.symbol.is_constant):
+            dimension_literal = dimension_ref.symbol.constant_value
+        # else exception is handled by the validate method.
+
+        # Determine the dimension and extent of the array
+        ndims = None
+        if len(array_ref.children) == 0:
+            # Note, the potential 'if not array_ref.symbol.is_array:'
+            # exception is already handled by the validate method.
+            ndims = len(array_ref.symbol.shape)
+
+            loop_bounds = []
+            for idx, shape in enumerate(array_ref.symbol.shape):
+                if shape in [ArrayType.Extent.DEFERRED,
+                             ArrayType.Extent.ATTRIBUTE]:
+                    # runtime extent using LBOUND and UBOUND required
+                    lbound = BinaryOperation.create(
+                        BinaryOperation.Operator.LBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    ubound = BinaryOperation.create(
+                        BinaryOperation.Operator.UBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    loop_bounds.append((lbound, ubound))
+                elif isinstance(shape, ArrayType.ArrayBounds):
+                    # array extent is defined in the array declaration
+                    loop_bounds.append(shape)
+                #  Note, the validate method guarantees that an else
+                #  clause is not required.
+        else:
+            # The validate method guarantees that this is an array
+            # reference.
+            loop_bounds = []
+            ndims = len(array_ref.children)
+            for shape in array_ref.children:
+                loop_bounds.append((shape.start.copy(), shape.stop.copy()))
+
+        # Determine the datatype of the array's values and create a
+        # scalar of that type
+        array_intrinsic = array_ref.symbol.datatype.intrinsic
+        array_precision = array_ref.symbol.datatype.precision
+        scalar_type = ScalarType(array_intrinsic, array_precision)
+
+        symbol_table = node.scope.symbol_table
+        assignment = node.ancestor(Assignment)
```

Comment:
Actually, the implementation works if `sum`, `maxval` or `minval` appear on the LHS. I've added a test in `mms_base_trans_test` called `test_lhs` to check. Perhaps I should have a separate test for the subclasses, rather than a single one in the base class?

Reply:
Ah, good spot again. No, there is an assumption that the intrinsic is part of an assignment. I've tightened the validation to raise an exception if this is not the case.